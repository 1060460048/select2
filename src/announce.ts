export function initialize() {
    if (document.getElementById("s2-live")) {
        return;
    }

    const live = document.createElement("div");
    live.setAttribute("id", "s2-live");
    live.setAttribute("class", "s2-offscreen s3-live");
    document.body.appendChild(live);

    const assertive = document.createElement("div");
    assertive.setAttribute("id", "s2-live-assertive");
    assertive.setAttribute("role", "log");
    assertive.setAttribute("aria-live", "assertive");
    assertive.setAttribute("aria-relevant", "additions");
    live.appendChild(assertive);

    const polite = document.createElement("div");
    polite.setAttribute("id", "s2-live-polite");
    polite.setAttribute("role", "log");
    polite.setAttribute("aria-live", "polite");
    polite.setAttribute("aria-relevant", "additions");
    live.appendChild(polite);

}

export function assertively(message: string) {
    const node = document.createElement("div");
    node.appendChild(document.createTextNode(message));
    document.getElementById("s2-live-assertive").appendChild(node);
    // TODO clean up old nodes


    // for debugging
    const live = document.getElementById("s2-live");
    live.scrollTop = live.scrollHeight;

}

export function politely(message: string) {
    const node = document.createElement("div");
    node.appendChild(document.createTextNode(message));
    document.getElementById("s2-live-polite").appendChild(node);


    // for debugging
    const live = document.getElementById("s2-live");
    live.scrollTop = live.scrollHeight;

    // TODO clean up old nodes
}
