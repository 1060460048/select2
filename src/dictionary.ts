export interface Dictionary {
    valueAdded(itemLabel: string): string;

    noSearchResults(): string;

    searchResultsLoading(): string;

    removeButtonTitle(): string;

    minimumCharactersMessage(len: number, min: number): string;

    instructions(): string;
}

class AmericanEnglishDictionary implements Dictionary {

    noSearchResults() {
        return "No results available";
    }

    searchResultsLoading() {
        return "Loading...";
    }

    removeButtonTitle() {
        return "Remove selected values";
    }

    valueAdded(itemLabel: string) {
        return itemLabel + " added";
    }

    minimumCharactersMessage(len: number, min: number) {
        const delta = min - len;
        return "Please enter " + delta + " more character" + (delta > 1 ? "s" : "");
    }

    instructions(): string {
        return "Items can be removed from this list box by selecting them and activating 'Remove selected values' button. Items can be added by selecting them in the adjacent combobox.";
    }
}

const dictionaries = new Map<string, Dictionary>();
dictionaries.set("en_us", new AmericanEnglishDictionary());

export function dictionary(dict: (Dictionary | string) = undefined): Dictionary {
    const fallback = dictionaries.get("en_us");

    if (!dict) {
        return fallback;
    }

    if (typeof dict == "string") {
        const instance = dictionaries.get(dict);
        return instance ? instance : fallback;
    } else {
        return dict;
    }
}
