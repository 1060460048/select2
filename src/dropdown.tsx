import * as React from "react";
import * as util from "./util";
import {createPortal} from "react-dom";

export interface DropdownProps {
    control: HTMLElement, // TODO make this a ref
    className?: string,
    children: React.ReactNode
}

export class Dropdown extends React.PureComponent<DropdownProps, {}> {

    container: HTMLElement;

    scrollParents: EventTarget[];
    private readonly throttledPosition: () => void;

    constructor(props: DropdownProps) {
        super(props);
        this.throttledPosition = util.throttle(50, this.position);
    }

    componentWillMount() {
        this.container = document.createElement("div");
        const cn = this.props.className;
        if (cn) {
            this.container.setAttribute("class", cn);
        }
        document.body.appendChild(this.container);

    }

    componentWillUnmount() {
        this.scrollParents.forEach((parent) => {
            ["resize", "scroll", "touchmove"].forEach((event) => {
                parent.removeEventListener(event, this.throttledPosition);
            });
        });
        delete this.scrollParents;
        document.body.removeChild(this.container);
        delete this.container;
    }


    render() {
        return createPortal(this.props.children, this.container);
    }

    componentDidMount() {
        this.scrollParents = util.getScrollParents(this.props.control);
        this.scrollParents.forEach((parent) => {
            ["resize", "scroll", "touchmove"].forEach((event) => {
                parent.addEventListener(event, this.throttledPosition);
            });
        });
        this.position();
    }

    componentDidUpdate() {
        this.position();
    }

    position = () => {
        const control = this.props.control;
        const rect = control.getBoundingClientRect();
        const style = `
            top: ${rect.top + rect.height + window.pageYOffset}px;
            left: ${rect.left + window.pageXOffset}px;
            width: ${rect.width}px;
        `;
        // TODO also called on scroll of children - should not be
        // console.log("position");
        this.container.setAttribute("style", style);
    }
}

