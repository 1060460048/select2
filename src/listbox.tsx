import * as React from "react";
import {ReactChild} from "react";
import {Key, uuid} from "./util";

interface ListboxProps<T> {
    items: T[];
    selected: number[];

    onChange(selected: number[]): void;

    elementRef?: React.RefObject<HTMLDivElement>;

    itemClassName(item: T, index: number, selected: boolean, active: boolean): string;

    itemLabel: ((item: T) => string) | (keyof T);
    itemContent?: ((item: T) => (string | React.ReactChild)) | (keyof T);

    orientation?: "horizontal" | "vertical";
    ariaLabel?: string;
    ariaLabelledBy?: string;
    ariaDescribedBy?: string;
    className?: string;

    onFocus?(event: React.FocusEvent): void;

    onBlur?(event: React.FocusEvent): void;
}

interface ListboxState<T> {
    active: number;
}

export class Listbox<T> extends React.PureComponent<ListboxProps<T>, ListboxState<T>> {

    static defaultProps = {
        orientation: "vertical"
    };

    id: string;
    ref: React.RefObject<HTMLDivElement>;

    constructor(props: ListboxProps<T>) {
        super(props);

        this.ref = React.createRef();

        this.state = {
            active: -1,
        }
    }

    componentWillMount() {
        this.id = uuid();
    }


    focus() {
        this.ref.current.focus();
    }

    render() {
        const props = this.props;
        const state = this.state;

        return (
            <div ref={this.ref}
                 className={props.className}
                 tabIndex={0}
                 role="listbox"
                 aria-orientation={props.orientation}
                 aria-activedescendant={this.getValueDomId(state.active)}
                 aria-multiselectable="true"
                 aria-label={props.ariaLabel}
                 aria-labelledby={props.ariaLabelledBy}
                 aria-describedby={props.ariaDescribedBy}
                 onKeyDown={this.onKeyDown}
                 onFocus={this.onFocus}
                 onBlur={this.onBlur}>
                {props.items.map((value, index) => {
                    const selected = props.selected.indexOf(index) >= 0;
                    const active = state.active == index;
                    const id = this.getValueDomId(index);
                    const classes = props.itemClassName(value, index, selected, active)
                    return (
                        <div key={index}
                             id={id}
                             className={classes}
                             role="option"
                             aria-selected={selected}
                             aria-checked={selected}
                             aria-label={this.getItemLabel(value)}
                             onClick={this.onItemClicked(index)}>
                            {this.getItemContent(value)}
                        </div>
                    );
                })}
            </div>
        );
    }

    onItemClicked = (index: number) => (event: React.MouseEvent) => {
        this.toggleSelectedValue(index);
        this.ref.current.focus();
        event.stopPropagation(); // do not propagate to body listener
    };


    getValueDomId(index: number) {
        return (index === undefined || index < 0) ? undefined : this.id + "-value-" + index;
    }

    onKeyDown = (event: React.KeyboardEvent) => {
        const values = this.props.items;
        const active = this.state.active;

        //console.log(event.key);

        switch (event.key) {
            case Key.ArrowLeft:
            case Key.ArrowUp: {
                if (active > 0) {
                    this.setActiveValue(active - 1);
                }
                event.preventDefault();
                break;
            }
            case Key.ArrowRight:
            case Key.ArrowDown: {
                if (active < values.length - 1) {
                    this.setActiveValue(active + 1);
                }
                event.preventDefault();
                break;
            }
            case Key.PageDown: {
                // TODO
                event.preventDefault();
                break;
            }
            case Key.PageUp: {
                // TODO
                event.preventDefault();
                break;
            }
            case Key.Home: {
                this.setActiveValue(0);
                event.preventDefault();
                break;
            }
            case Key.End: {
                this.setActiveValue(values.length - 1);
                event.preventDefault();
                break;
            }
            case Key.Space: {
                this.toggleSelectedValue(active);
                event.preventDefault();
                break;
            }
        }
    };

    onFocus = (event: React.FocusEvent) => {
        if (this.state.active < 0) {
            if (this.props.items.length > 0) {
                let index = 0;
                if (this.props.selected.length > 0) {
                    index = this.props.selected[0];
                }
                this.setActiveValue(index);
            }
        }
        if (this.props.onFocus) {
            this.props.onFocus(event);
        }
    };

    onBlur = (event: React.FocusEvent) => {
        this.setActiveValue(-1);
        if (this.props.onBlur) {
            this.props.onBlur(event);
        }
    };


    setActiveValue(index: number) {
        this.setState({active: index});
    }

    toggleSelectedValue(index: number) {
        const oldSelected = this.props.selected;
        const newSelected = oldSelected.slice();
        const pos = oldSelected.indexOf(index);
        if (pos >= 0) {
            newSelected.splice(pos, 1);
        } else {
            newSelected.push(index);
        }

        this.props.onChange(newSelected);
    }

    getItemLabel(item: T): string {
        if (typeof this.props.itemLabel === "function") {
            return this.props.itemLabel(item);
        } else {
            return "" + item[this.props.itemLabel];
        }
    }

    getItemContent(item: T): ReactChild {
        if (this.props.itemContent === undefined) {
            return this.getItemLabel(item);
        } else if (typeof this.props.itemContent === "function") {
            return this.props.itemContent(item);
        } else {
            return "" + item[this.props.itemContent];
        }
    }


}