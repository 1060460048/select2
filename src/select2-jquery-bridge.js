;(function ($, window, document, React, ReactDOM, select2, undefined) {

    "use strict";

    var pluginName = "select2",
        defaults = {
            //propertyName: "value"
        };

    function Plugin(element, options) {
        this.element = element;

        this.settings = $.extend({}, defaults, options);
        this._defaults = defaults;
        this._name = pluginName;

        if (!this.settings.name) {
            this.settings.name = $(element).attr("name");
        }

        this.init();
    }

    $.extend(Plugin.prototype, {
        init: function () {

            var container = document.createElement("div");

            var target = $(this.element);
            target.hide();
            target.after(container);

            ReactDOM.render(React.createElement(select2.MultiSelect, this.settings), container);
        },

    });

    $.fn[pluginName] = function (options) {
        return this.each(function () {
            if (!$.data(this, "plugin_" + pluginName)) {
                $.data(this, "plugin_" +
                    pluginName, new Plugin(this, options));
            }
        });
    };

})(jQuery, window, document, React, ReactDOM, select2);