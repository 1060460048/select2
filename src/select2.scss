
$color-focus: #4D90FE;
$color-item-bkg: #0073E6;
$color-item-text: #fff;
$border: 1px solid #ddd;
$border-focus: 1px solid $color-focus;
$border-radius: 5px;
$dropdown-padding: 8px 10px;

/*
 content: "✓";
content: "×";
*/

.s25-control {
  position: relative;
  box-sizing: border-box;
  min-height: 36px;
  width: 100%;

}

.s25-multi {
  display: flex;
  flex-wrap: nowrap;
}

.s25-multi .s25-body {
  flex-grow: 1;
  display: flex;
  flex-wrap: wrap;
}

.s25-multi .s25-multi-values {
  box-sizing: border-box;
  display: flex;
  flex-wrap: wrap;
  
	.s25-item {
		cursor: pointer;
		box-sizing: border-box;
		align-items: center;
	}
}

.s25-multi div {
  max-width: 100%;
}

.s25-multi .s25-search {
  max-width: 100%;
  min-width: 75px;
  border: none;
}

.s25-multi .s25-toggle {
  display: flex;
  box-sizing: border-box;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 30px;
}

.s25-multi .s25-remove {
  cursor: pointer;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  
	&:disabled {
		cursor: default;
	}
	
	&:active {
		border: 0px;
	}
  
}

.s25-offscreen {
  clip-path: inset(100%);
  clip: rect(1px, 1px, 1px, 1px);
  width: 1px;
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
}

.s25-dropdown {
  position: absolute;
  z-index: 1000000;
  box-sizing: border-box;
  
	.s25-search-results {
		width: 100%;
		overflow-y: auto;
		position: relative;
		box-sizing: border-box;
	}
	
	.s25-options {
		width: 100%;
		
		.s25-item {
			display: block;
			width: 100%;
			box-sizing: border-box;
		}
	}
}

.s25-no-search-results {
  padding: $dropdown-padding;

}

.s25-search-results-message {
  padding: $dropdown-padding;
}

.s25-search-results-loading {
  /*! keep me */
}

.s25-search-results-minimum-error {
  /*! keep me */
}

.s25-search-results {
  /*! keep me */

}

.s25-open {
  /*! keep me */

}

.s25-search-container {
  /*! keep me */
}

/* skin */

.s25-control {
  border: $border;
  background-color: #fff;
  border-radius: $border-radius;

	&.s25-focused {
		border: $border-focus;
	
		&.s25-open {
			border-bottom: none;
			border-radius: $border-radius $border-radius 0 0;
		}
	}
}

.s25-control .s25-body {
  margin: 3px;
  margin-right: 60px;
  width: calc(100% - 60px);
}

.s25-search:focus {
  outline: 0;
}

.s25-multi .s25-toggle {
  width: 30px;
}

.s25-toggle > svg {
  display: inline-block;
  margin: auto;
}

.s25-multi {
  /*! keep me */

}

.s25-multi .s25-search {
  font-size: 100%;
  line-height: 24px;
}

.s25-multi-values {
  padding: 0;
}

.s25-content {
  /*! keep me */
}

.s25-multi-values:focus {
  outline: 0
}

.s25-multi-values .s25-item {
  position: relative;
  padding: 5px 12px 5px 18px;
  border-radius: $border-radius;
  background-color: $color-item-bkg;
  color: $color-item-text;
  margin-right: 3px;
  margin-bottom: 3px;
  transition: background .2s;
  
  	&:hover {
	  	background-color: darken( $color-item-bkg, 5% );
  	}
  
	&.s25-active {
		box-shadow: inset 0 0 0 3px #67B3FF;
	}

	&.s25-item.s25-selected {
		text-align: right;
		
		&:before {
			content: "✓";
			position: absolute;
			left: 5px;
		}
	}
}

.s25-multi-values .s25-label {
  color: rgb(34, 34, 34);
  font-size: 85%;
  border-radius: $border-radius;
  padding: 3px 3px 3px 6px;
}

.s25-multi .s25-body .s25-remove {
  border: none;
  border-radius: 0 $border-radius $border-radius 0;
  margin: 0;
  padding: 0 5px;
  outline: none;
  background-color: #f1f1f1;
  transition: background .2s;
	
	span svg {
		fill: #ccc;
		transition: fill .2s;
	}
	
	&:not(:disabled) {
		background-color: lighten( $color-item-bkg, 50% );
	
		&:hover {
		  	background-color: lighten( $color-item-bkg, 45% );
		  	
	  	}
	  	
  		span svg {
			fill: $color-item-bkg;
		}	
	}	
		
	&:focus {
		border: $border-focus;
	}
	
}

.s25-dropdown {
  border: $border-focus;
  border-top: $border;
  border-radius: 0 0 $border-radius $border-radius;
  background-color: rgb(255, 255, 255);


	.s25-body {
		padding-top: 4px;
		padding-bottom: 4px;
	}
}

.s25-live {
  border-radius: $border-radius;
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  min-height: 40px;
  max-height: 50px;
  border: 1px solid purple;
  overflow: scroll;
}

.s25-dropdown .s25-options {

	.s25-item {
		background-color: transparent;
		color: inherit;
		font-size: inherit;
		padding: $dropdown-padding;
	}

	.s25-active {
		background-color: lighten( $color-item-bkg, 50% );
	}
}

