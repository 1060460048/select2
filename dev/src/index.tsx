import * as React from "react";
import * as ReactDOM from "react-dom";

import countries from "./countries";
import * as select2 from "../../src/select2";
import {MultiSelect} from "../../src/select2";

type Country = { code: string, name: string };

const query = (function () {
    let previousTimeout: number = undefined;
    let previousReject: (reason: any) => any = undefined;

    return function (search: string, page: number) {
        if (previousReject != undefined) {
            previousReject(new Error("cancelled"));
            window.clearTimeout(previousTimeout);
        }
        const delay = 100;

        return new Promise<select2.QueryResult<Country>>((resolve: any, reject: any) => {
            previousReject = reject;
            previousTimeout = window.setTimeout(function () {
                const results: Country[] = [];
                let count = 0;
                let limit = 5;
                let offset = page * limit;
                for (let i = 0; i < countries.length; i++) {
                    const country = countries[i];
                    if (country.name.toLowerCase().indexOf(search.toLowerCase()) >= 0) {
                        if (count >= offset) {
                            results.push(country);
                        }
                        count++;
                        if (count > offset + limit) {
                            break;
                        }
                    }
                }
                previousTimeout = undefined;
                previousReject = undefined;
                resolve({values: results, more: results.length == limit})
            }, delay);
        });
    }
})();

ReactDOM.render(
    <MultiSelect
        name="countries"
        values={countries.splice(0, 3)}
        itemId="code"
        itemLabel={(c) => c.name}
        valueContent={(item)=><div style={{color:"green"}}>{item.name}</div>}
        resultContent={(item)=><div style={{color:"blue"}}>{item.name}</div>}
        valuesLabel="Countries"
        searchLabel="Add Country"
        query={query}
        minimumCharacters={2}
        openOnSearchFocus={true}
        quietMillis={100}

    />
    , document.getElementById("multi1")
);

ReactDOM.render(
    <MultiSelect
        name="countries2"
        values={countries.splice(0, 2)}
        itemId={(c) => c.code}
        itemLabel={(c) => c.name}
        valueContent={(c) => c.name}

        valuesLabel="Countries"
        searchLabel="Add Country"
        query={query}
        openOnSearchFocus={false}
    />
    , document.getElementById("multi2")
);
