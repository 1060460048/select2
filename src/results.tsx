import * as React from "react";
import * as style from "./select2.scss";
import * as classNames from "classnames/bind";
import {Dictionary} from "./dictionary";
import * as util from "./util";

let cn = classNames.bind(style);

export interface ResultsProps<T> {
    search: string;
    minimumCharacters: number;

    searchResults: T[] | undefined;
    searchResultsLoading: boolean;
    searchHasMoreResults: boolean;
    activeSearchResult: number;
    showNoSearchResultsFound: boolean;
    showLoadMoreResults: boolean;
    showMinimumCharactersError: boolean;

    listboxDomId: string;

    searchResultDomId(index:number):string;

    itemLabel(item: T): string;

    resultContent(item: T): (string | React.ReactChild);

    dictionary: Dictionary;

    onLoadMore(): void;

    onActivateSearchResult(index: number): void;

    onSelectSearchResult(index: number): void;
}

export class Results<T> extends React.PureComponent<ResultsProps<T>, {}> {

    containerRef: React.RefObject<HTMLDivElement>;
    loadingMoreResults: React.RefObject<HTMLDivElement>;

    lastMouseClientX: number;
    lastMouseClientY: number;


    constructor(props: ResultsProps<T>) {
        super(props);
        this.containerRef = React.createRef();
        this.loadingMoreResults = React.createRef();
    }

    render() {

        const props = this.props;
        const dict = props.dictionary;

        return <div className={style.s25SearchResults}
                    style={{maxHeight: "120px"}}
                    ref={this.containerRef}
                    aria-busy={props.searchResultsLoading}
                    onScroll={this.onScroll}>
            {props.searchResults && props.searchResults.length > 0 &&
            (
                <div
                    id={props.listboxDomId}
                    className={style.s25Options}
                    role="listbox"
                    aria-activedescendant={this.props.searchResultDomId(props.activeSearchResult)}>
                    {
                        props.searchResults.map((item, i) => {
                            const className = cn(style.s25Item, {s25Active: props.activeSearchResult == i});
                            return (
                                <div key={i}
                                     className={className}
                                     id={this.props.searchResultDomId(i)}
                                     onMouseDown={this.onMouseDownSearchResult(i)}
                                     onMouseMove={this.onMouseEnterSearchResult(i)}
                                     onMouseUp={this.onSearchResultClicked(i)}
                                     role="option"
                                     aria-posinset={i + 1}
                                     aria-selected={props.activeSearchResult === i}
                                     aria-label={props.itemLabel(item)}
                                >
                                    {props.resultContent(item)}
                                </div>);
                        })
                    }
                </div>

            )}
            {props.showNoSearchResultsFound &&
            <div className={style.s25NoSearchResults}>{dict.noSearchResults()}</div>
            }
            {(props.searchResultsLoading || props.showLoadMoreResults) &&
            <div ref={this.loadingMoreResults}
                 className={cn(style.s25SearchResultsLoading, style.s25SearchResultsMessage)}>{dict.searchResultsLoading()}</div>
            }
            {(props.showMinimumCharactersError) &&
            <div className={cn(style.s25SearchResultsMinimumError, style.s25SearchResultsMessage)}>
                {dict.minimumCharactersMessage(props.search.length, props.minimumCharacters)}
            </div>
            }
        </div>
    }

    onScroll = (event: React.UIEvent) => {
        if (!this.props.showLoadMoreResults) {
            return;
        }
        const more = this.loadingMoreResults.current;
        const drop = this.containerRef.current;

        const visibility = util.calculateVerticalVisibility(drop, more);
        if (visibility !== "hidden") {
            this.props.onLoadMore();
        }

    };


    onMouseDownSearchResult = (index: number) => (event: React.MouseEvent) => {
        if (this.props.activeSearchResult != index) {
            this.props.onActivateSearchResult(index);
        }
        event.preventDefault();
    };

    onMouseEnterSearchResult = (index: number) => (event: React.MouseEvent) => {
        if (this.lastMouseClientX == event.clientX && this.lastMouseClientY == event.clientY) {
            // the mouse did not move, the dropdown was scrolled instead, we do not change selected element because
            // it will be scrolled into view and mess with the scrolling of the results in the dropdown
            return;
        }


        this.lastMouseClientX = event.clientX;
        this.lastMouseClientY = event.clientY;
        if (this.props.activeSearchResult != index) {
            this.props.onActivateSearchResult(index);
        }
    };

    onSearchResultClicked = (index: number) => (event: React.MouseEvent) => {
        this.props.onSelectSearchResult(index);
        event.preventDefault();
    };

    componentDidUpdate(prevProps: ResultsProps<T>, prevState: ResultsProps<T>) {

        if (prevProps.activeSearchResult != this.props.activeSearchResult) {

            if (this.props.activeSearchResult >= 0
                && this.props.searchResults.length > 0
                && this.props.activeSearchResult == (this.props.searchResults.length - 1)
                && this.props.showLoadMoreResults) {
                // last result is selected and load more is shown, make sure it is scrolled into view

                const drop = this.containerRef.current;
                const el = this.loadingMoreResults.current;

                drop.scrollTop = el.offsetTop + el.offsetHeight - drop.clientHeight;


                //console.log("scrolling to see load more");//, setting scrolltop", drop, el, el.offsetTop - drop.clientHeight);


            } else if (this.props.activeSearchResult >= 0) {
                // make sure it is scrolled into view
                const id = this.props.searchResultDomId(this.props.activeSearchResult);
                const el = document.getElementById(id);
                if (el != null) {
                    const drop = this.containerRef.current;

                    const c = drop.getBoundingClientRect();
                    const e = el.getBoundingClientRect();

                    if (e.top < c.top && e.bottom <= c.bottom) {
                        const delta = c.top - e.top;
                        drop.scrollTop = drop.scrollTop - delta;
                        //console.log("scrolling into view top");
                    }

                    if (e.top >= c.top && e.bottom > c.bottom) {
                        const delta = e.bottom - c.bottom;
                        drop.scrollTop = drop.scrollTop + delta;
                        //console.log("scrolling into view bottom");
                    }

                }
            }
        }
    }
}