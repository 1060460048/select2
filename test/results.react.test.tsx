import * as React from "react";
import {Results} from "../src/results";
import {ReactWrapper, shallow, ShallowWrapper} from "enzyme";
import toJson from "enzyme-to-json";
import {makeQuery, waitForState} from "./utils";
import {dictionary} from "../src/dictionary";


test("No Results", async () => {
    let component = shallow(
        <Results
            showNoSearchResultsFound={true}

            search={"test"}
            minimumCharacters={3}
            searchResults={undefined}
            searchResultDomId={(index)=>"result-"+index}
            searchHasMoreResults={false}
            activeSearchResult={undefined}
            searchResultsLoading={false}
            showLoadMoreResults={false}
            showMinimumCharactersError={false}
            listboxDomId={"listbox"}
            itemLabel={undefined}
            resultContent={undefined}
            dictionary={dictionary()}
            onLoadMore={undefined}
            onActivateSearchResult={undefined}
            onSelectSearchResult={undefined}
        />);

    expect(toJson(component)).toMatchSnapshot();
});

test("Minimum Characters For Search", async ()=> {

    let component = shallow(
        <Results
            search={""}
            minimumCharacters={2}
            showMinimumCharactersError={true}

            showNoSearchResultsFound={false}
            searchResults={undefined}
            searchResultDomId={(index)=>"result-"+index}
            searchHasMoreResults={false}
            activeSearchResult={undefined}
            searchResultsLoading={false}
            showLoadMoreResults={false}
            listboxDomId={"listbox"}
            itemLabel={undefined}
            resultContent={undefined}
            dictionary={dictionary()}
            onLoadMore={undefined}
            onActivateSearchResult={undefined}
            onSelectSearchResult={undefined}
        />);

    expect(toJson(component)).toMatchSnapshot("enter two more");

    component = shallow(
        <Results
            search={"t"}
            minimumCharacters={2}
            showMinimumCharactersError={true}

            showNoSearchResultsFound={false}
            searchResults={undefined}
            searchResultDomId={(index)=>"result-"+index}
            searchHasMoreResults={false}
            activeSearchResult={undefined}
            searchResultsLoading={false}
            showLoadMoreResults={false}
            listboxDomId={"listbox"}
            itemLabel={undefined}
            resultContent={undefined}
            dictionary={dictionary()}
            onLoadMore={undefined}
            onActivateSearchResult={undefined}
            onSelectSearchResult={undefined}
        />);

    expect(toJson(component)).toMatchSnapshot("enter one more");

    component = shallow(
        <Results
            search={"te"}
            minimumCharacters={2}
            showMinimumCharactersError={false}

            showNoSearchResultsFound={false}
            searchResults={[{id:1,text:"item-1"},{id:2,text:"item-2"}]}
            searchResultDomId={(index)=>"result-"+index}
            searchHasMoreResults={false}
            activeSearchResult={undefined}
            searchResultsLoading={false}
            showLoadMoreResults={false}
            listboxDomId={"listbox"}
            itemLabel={(i)=>i.text}
            resultContent={(i)=>i.text}
            dictionary={dictionary()}
            onLoadMore={undefined}
            onActivateSearchResult={undefined}
            onSelectSearchResult={undefined}
        />);

    expect(toJson(component)).toMatchSnapshot("results");

});
