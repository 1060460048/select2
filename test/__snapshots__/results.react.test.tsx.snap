// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Minimum Characters For Search: enter one more 1`] = `
<div
  aria-busy={false}
  className="s25SearchResults"
  onScroll={[Function]}
  style={
    Object {
      "maxHeight": "120px",
    }
  }
>
  <div
    className="s25SearchResultsMinimumError s25SearchResultsMessage"
  >
    Please enter 1 more character
  </div>
</div>
`;

exports[`Minimum Characters For Search: enter two more 1`] = `
<div
  aria-busy={false}
  className="s25SearchResults"
  onScroll={[Function]}
  style={
    Object {
      "maxHeight": "120px",
    }
  }
>
  <div
    className="s25SearchResultsMinimumError s25SearchResultsMessage"
  >
    Please enter 2 more characters
  </div>
</div>
`;

exports[`Minimum Characters For Search: results 1`] = `
<div
  aria-busy={false}
  className="s25SearchResults"
  onScroll={[Function]}
  style={
    Object {
      "maxHeight": "120px",
    }
  }
>
  <div
    aria-activedescendant="result-undefined"
    className="s25Options"
    id="listbox"
    role="listbox"
  >
    <div
      aria-label="item-1"
      aria-posinset={1}
      aria-selected={false}
      className="s25Item"
      id="result-0"
      key="0"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      role="option"
    >
      item-1
    </div>
    <div
      aria-label="item-2"
      aria-posinset={2}
      aria-selected={false}
      className="s25Item"
      id="result-1"
      key="1"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      role="option"
    >
      item-2
    </div>
  </div>
</div>
`;

exports[`No Results 1`] = `
<div
  aria-busy={false}
  className="s25SearchResults"
  onScroll={[Function]}
  style={
    Object {
      "maxHeight": "120px",
    }
  }
>
  <div
    className="s25NoSearchResults"
  >
    No results available
  </div>
</div>
`;
